libsolidity/analysis/TypeChecker.cpp

```diff
@@ -523,6 +538,10 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 			m_errorReporter.typeError(6744_error, _variable.location(), "Internal or recursive type is not allowed for public state variables.");
 	}
 
+	bool isStructMemberDeclaration = dynamic_cast<StructDefinition const*>(_variable.scope()) != nullptr;
+	if (isStructMemberDeclaration)
+		return false;
```

Comment:
Is it an appropriate way to filter struct members?

Reply:
Yes, I was also thinking if we should test types themselves, or just variables. We could report errors for bad-for-any-location types too.

The second issue here is that (for historical reasons?) we report collision warnings for state variables only, https://github.com/ethereum/solidity/pull/9062/files#diff-9f2a45efaa7b75ab7ec620d3177cf01aR558.

So for now, both `contract C` and `contract D` produce no collision warnings.

Reply:
We can think about warning for ``C`` at a later point. For ``D`` it makes sense to have the warning, though - the suggestion below about the "if state variable" part might fix that, though.

Reply:
`D` is added to tests.